keps/sig-auth/20191204-x509-certificate-for-serviceaccount.md

```diff
@@ -0,0 +1,286 @@
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP describes Kubernetes generates and mounts ServiceAccounts' x509 certificates for Pods, 
+and Kubernetes can get ServiceAccount identity information if Pods visit Kubernetes using ServiceAccounts' x509 certificates.
+
+## Background
+
+By default, Pod running in Kubernetes will be mount Kubernetes root CA and ServiceAccount JWT(which will be mount at path `/var/run/secrets/kubernetes.io/serviceaccount`). 
+And Kubernetes can get ServiceAccount identity information if Pod visit Kubernetes using SerivceAccount JWT(common usage: `client-go`), 
+then Kubernetes will do Authz according the ServiceAccount identity information and the requested resources.
+
+## Motivation
+
+In a large scale Kubernetes cluster, we found Kubernetes will meet performance issue if Client send too many requests to Kubernetes using JWT as their identity. 
+As Kubernetes must call etcd once to check Token is valid in serious situation, the serious situation including: 
+JWT dose not hit token result cache and Secret informer, ServiceAccountToken even does not exist or had been deleted, etc. 
+More discuss please see [#83259](https://github.com/kubernetes/kubernetes/issues/83259), and for now, we can not found a solution to solve serious situation but increase the cache size, 
```

Comment:
This wasn't how we fixed this issue.

https://github.com/kubernetes/kubernetes/pull/71816
https://github.com/kubernetes/kubernetes/pull/79375
https://github.com/kubernetes/kubernetes/pull/80465
https://github.com/kubernetes/kubernetes/pull/84424
https://github.com/kubernetes/kubernetes/pull/85242

Reply:
These PRs seems not solve the token lookup failure situation. 

What if 10k times visit API Server using a not-exist SA Token at same time.

Reply:
That's what failure TTL is for.

Reply:
I do not think we can set failure TTL to be a positive value. 

For example, we set it as 10 seconds, and the ServiceAccount Secret is generated by the controller, it may created at the 1st seconds, if a request using this ServiceAccount reached before ServiceAccount Secret created, for example the 0th second, then this token would be invalid before 10th second.